crates/cairo-profiler/src/profile_viewer.rs

```diff
@@ -0,0 +1,168 @@
+            profile.string_table[usize::try_from(sample.unit)
+                .expect("Overflow while converting samples id to usize")]
+                == sample_label
+        })
+        .context("Failed to find sample in provided profile")?;
+
+    let mut profile_map = HashMap::<String, FunctionProfile>::new();
+
+    let location_map: HashMap<u64, &Location> = profile
+        .location
+        .iter()
+        .map(|location| (location.id, location))
+        .collect();
+
+    let function_map: HashMap<u64, &Function> = profile
+        .function
+        .iter()
+        .map(|function| (function.id, function))
+        .collect();
+
+    for sample in &profile.sample {
+        let sample_value = sample.value[sample_type_idx];
+
+        sample
+            .location_id
+            .iter()
+            .filter_map(|&loc_id| {
+                location_map
+                    .get(&loc_id)
+                    .and_then(|location| location.line.first())
+                    .and_then(|line| function_map.get(&line.function_id))
+                    .map(|function| {
+                        &profile.string_table[usize::try_from(function.name)
+                            .expect("Overflow while converting function id to usize")]
+                    })
+            })
+            .enumerate()
+            .for_each(|(idx, function_name)| {
+                let entry = profile_map.entry(function_name.clone()).or_default();
+                entry.cumulative += sample_value;
+                if idx == 0 {
+                    entry.flat += sample_value;
+                }
+            });
+    }
+
+    let total_resource_count = profile_map
+        .values()
+        .max_by_key(|function| function.cumulative)
+        .map(|function_profile| function_profile.cumulative)
+        .context("Failed to obtain total resource count from cumulative stats")?;
+
+    // sum_p depends on the correct order of data
+    let mut sorted_profile_map: Vec<(String, FunctionProfile)> = profile_map.into_iter().collect();
+    sorted_profile_map.sort_by(|a, b| b.1.flat.cmp(&a.1.flat));
+
+    let mut sum_p: f64 = 0.0;
+
+    for (_, profile) in &mut sorted_profile_map {
+        sum_p += profile.flat as f64;
+
+        profile.flat_p = (profile.flat as f64 / total_resource_count as f64) * 100.0;
+        profile.cumulative_p = (profile.cumulative as f64 / total_resource_count as f64) * 100.0;
+        profile.sum_p = (sum_p / total_resource_count as f64) * 100.0;
+    }
+
+    Ok(sorted_profile_map)
+}
+
+pub fn get_samples(profile: &Profile) -> Vec<String> {
+    profile
+        .sample_type
+        .iter()
+        .map(|sample| {
+            profile.string_table[usize::try_from(sample.unit)
+                .expect("Overflow while converting samples id to usize")]
+            .clone()
+        })
+        .collect()
+}
+
+pub fn print_profile(profile: &Profile, sample: &str, limit: usize) -> Result<()> {
+    if limit == 0usize {
```

Comment:
Instead of validating it manually, we can make limit a NonZeroUsize, which would allow Clap to automatically handle the validation. It also provides a hint to the reader, and being non-zero would be enforced at the type level

Reply:
![obraz](https://github.com/user-attachments/assets/e6f63a6c-d2d9-4aa5-b282-d1e1c22ebe6d)


didn't know NonZeroUsize was a thing, thanks!